File: 70. Climbing Stairs/Climbing_Stairs.js
//This is not a perfect solution for this problem
//You must learn DP
//See the CPP solution

const factorial = function (n) {
  if (n == 0) return 1;
  let factorial = 1;
  for (let i = 2; i <= n; i++) factorial = factorial * i;
  return factorial;
};

const nCr = function (n, r) {
  return factorial(n) / (factorial(r) * factorial(n - r));
};

var climbStairs = function (n) {
  let extra = 1;
  let r = 1,
    add = 0;
  if (n % 2 == 0) extra = 2;
  n = n - 1;
  while (n > r) {
    add = add + nCr(n, r);
    n = n - 1;
    r = r + 1;
  }
  add += extra;
  return add;
};

let result = climbStairs(45);
console.log(result);

File: 70. Climbing Stairs/Climbing_Stairs.cpp
#include <iostream>

using namespace std;

int climbStairs(int n) {
    if(n == 1 || n == 0) return 1;
    int ways[n+1];
    ways[0] = ways[1] = 1;
    for(int i=2; i<n+1; i++) {
        ways[i] = ways[i-1] + ways[i-2]; 
    }
    return ways[n];
}

//This recursive solution is not efficient. It get worse when n is a larger value
// int climbStairs(int n) {
//     if(n == 1 || n == 0) return 1;
//     return climbStairs(n-1) + climbStairs(n-2);
// }

int main() {
    int result = climbStairs(45);
    cout<<result;
    return 0;
}